dotnet/src/Azure.Iot.Operations.Protocol/CloudEvent-Architecture.md

```diff
@@ -0,0 +1,225 @@
+# CloudEvent Architecture
+
+This document describes the dual CloudEvent implementation in Azure IoT Operations Protocol SDK.
+
+## Overview
+
+The SDK provides two CloudEvent classes to handle different usage scenarios:
+
+1. **CloudEvent** - For generic MQTT scenarios where users control all properties
+2. **ProtocolCloudEvent** - For protocol-specific scenarios where the SDK controls Type and DataContentType
+
+## Class Diagram
+
+```mermaid
+classDiagram
+    class CloudEvent {
+        +Uri Source
+        +string SpecVersion
+        +string Type
+        +string? Id
+        +DateTime? Time
+        +string? DataContentType
+        +string? Subject
+        +string? DataSchema
+        +CloudEvent(source, type, specversion)
+    }
+    
+    class ProtocolCloudEvent {
+        +Uri Source
+        +string SpecVersion
+        +string Type
+        +string? Id
+        +DateTime? Time
+        +string? DataContentType
+        +string? Subject
+        +string? DataSchema
+        +ProtocolCloudEvent(source, specversion)
+    }
+    
+    class MqttApplicationMessage {
+        +SetCloudEvent(CloudEvent)
+        +CloudEvent? GetCloudEvent()
+        +SetCloudEvent(ProtocolCloudEvent) internal
+        +ProtocolCloudEvent? GetProtocolCloudEvent() internal
+    }
+    
+    class OutgoingTelemetryMetadata {
+        +ProtocolCloudEvent? CloudEvent
+        +Dictionary~string,string~ UserData
+    }
+    
+    class IncomingTelemetryMetadata {
+        +ProtocolCloudEvent GetCloudEvent()
```

Comment:
Should there be equivalent GetCloudEvent fns on the command metadatas, or is there a difference that makes it not needed?